// experian/index.js
var xmljson = require('xmljson');
var request = require('request');

function Autocheck(username, password, live) {
  
  if (live == true) {
    this.url = 'https://www.automotivemxin2.com/lambert';
  } else {
    this.url = 'https://uat.automotivemxin.com';
  }
  
  this.username = username;
  this.password = password;
}

Autocheck.prototype._build_xml = function(object) {
  var string = '';
  
  for (var field in object) {
    if (object.hasOwnProperty(field)) {

      var deeper = object[field];
      
      field = field.toUpperCase();
      
      if (typeof deeper == 'object' && deeper !== null) {
        string += '<' + field + '>\r\n' + this._build_xml(deeper) + '</' + field + '>\r\n';
      } else {
        string += '<' + field + '>' + deeper + '</' + field + '>\r\n';
      }
    }
  }
  
  return string;
};

Autocheck.prototype._build_request = function(options) {
  var req = {
    'experian': {
      'eseries': {
        'form_id': 'B2INT'
      },
      'mxin': {
        'username': this.username,
        'password': this.password,
        'paymentcollectiontype': '02'
      }
    }
  };
  
  for (var option in options) {
    if (options.hasOwnProperty(option)) {
      req.experian.mxin[option] = options[option];
    }
  }
  
  var xml = '<?xml version="1.0" encoding="utf-8"?>\r\n' + this._build_xml(req);
  return xml;
}

Autocheck.prototype.request = function(options, callback) {
  var req = this._build_request(options);
  request.post({
    uri: this.url,
    body: req
  }, function(err, res, body) {    
    var parser = new xmljson();
    
    if (err) {
      callback({message: err});
    } else {
      parser.on('end', function(result) {
        if (!!result.request.err1) {
          callback({message: result.request.err1.message});
        } else {
          callback(null, result);
        }
      });

      body = body.replace("<?xml version='1.0' standalone='yes' ?>", '');
    
      parser.parse(body);
      parser.close();
    }
  });
};

module.exports = Autocheck;
